osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+
+namespace osu.Game.Rulesets.Difficulty.Skills
+{
+    /// <summary>
+    /// Used to processes strain values of <see cref="DifficultyHitObject"/>s, keep track of strain levels caused by the processed objects
+    /// and to calculate a final difficulty value representing the difficulty of hitting all the processed objects.
+    /// </summary>
+    public abstract class StrainSkill : Skill
+    {
+        /// <summary>
+        /// The peak strain for each <see cref="SectionLength"/> section of the beatmap.
+        /// </summary>
+        public IReadOnlyList<double> StrainPeaks => strainPeaks;
+
+        /// <summary>
+        /// Strain values are multiplied by this number for the given skill. Used to balance the value of different skills between each other.
+        /// </summary>
+        protected abstract double SkillMultiplier { get; }
+
+        /// <summary>
+        /// Determines how quickly strain decays for the given skill.
+        /// For example a value of 0.15 indicates that strain decays to 15% of its original value in one second.
+        /// </summary>
+        protected abstract double StrainDecayBase { get; }
+
+        /// <summary>
+        /// The weight by which each strain value decays.
+        /// </summary>
+        protected virtual double DecayWeight => 0.9;
+
+        /// <summary>
+        /// The current strain level.
+        /// </summary>
+        protected double CurrentStrain { get; private set; } = 1;
+
+        /// <summary>
+        /// The length of each strain section.
+        /// </summary>
+        protected virtual int SectionLength => 400;
+
+        private double currentSectionPeak = 1; // We also keep track of the peak strain level in the current section.
+
+        private double currentSectionEnd;
+
+        private readonly List<double> strainPeaks = new List<double>();
+
+        /// <summary>
+        /// Process a <see cref="DifficultyHitObject"/> and update current strain values accordingly.
+        /// </summary>
+        public sealed override void Process(DifficultyHitObject current)
+        {
+            // The first object doesn't generate a strain, so we begin with an incremented section end
+            if (Previous.Count == 0)
+                currentSectionEnd = Math.Ceiling(current.StartTime / SectionLength) * SectionLength;
```

Comment:
I am *relatively* certain that moving this is the cause of the slight taiko diffcalc shift. If you look at all rulesets' implementations of `CreateDifficultyHitObjects()`, they don't generate a difficulty hitobject for the first object of the map, because doing so usually requires at least one preceding object to exist, so they usually start from the second or even third one. That *may* be fine to do, but it is a behavioural change.

Reply:
Ah that could definitely be it! I don't think that is enough of a reason to require strain calculations to happen at a `DifficultyCalculator` level though. If we want to make certain that this tiny output change doesn't occur we can find another way around it.

Reply:
In fact thinking about this again, doesn't this logically make more sense than before?
If a calculator doesn't create difficulty hitobjects for the first one or two hitobjects, then that carries the implication that those first objects "don't matter for difficulty".
With that implication, then it would make sense for the strain decay to begin at the first object that does contribute to difficulty, rather than letting it decay from the starting value 1 to some value slightly lower before real difficulty calculation begins (which is what happened before, with it starting at the first REAL hitobjects time).

My preference would be to actually enforce 1:1 hitobject to difficulty hitobject (excluding spinners for example) transformation and just handle the cases for missing previous objects within the difficulty code... but that's a discussion for another time.

Reply:
Sure, I'm not expecting my original remark to be "resolved" in any capacity, per se. Was more of an observation if anything.